File: slidingWindowMedian/slidingWindowMedian.cpp
// clang++ slidingWindowMedian.cpp -std=c++11

#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Solution {
public:
    vector<double> medianSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        vector<double> res(n - k + 1);
        for (int i = 0; i < nums.size(); i++) {
            if (i >= k) {
                // remove at first
                if (leftS.count(nums[i - k]))
                    leftS.erase(leftS.find(nums[i - k]));
                else
                    rightS.erase(rightS.find(nums[i - k]));
                // adjust to properity
                if (leftS.size() - rightS.size() == 2) {
                    rightS.insert(*leftS.rbegin());
                    leftS.erase(--leftS.end());
                } else if (leftS.size() + 1 == rightS.size()) {
                    leftS.insert(*rightS.begin());
                    rightS.erase(rightS.begin());
                }
            }
            // add
            addNum(nums[i]);
            // getMedian
            if (i + 1 >= k)
                res[i - k + 1] = getMedian();
        }
        return res;
    }
    void addNum(int num) {
        leftS.insert(num);
        rightS.insert(*leftS.rbegin());
        leftS.erase(--leftS.end());
        if (leftS.size() < rightS.size()) {
            leftS.insert(*rightS.begin());
            rightS.erase(rightS.begin());
        }
    }
    double getMedian() {
        if (leftS.empty())
            return -1.0;
        if (leftS.size() > rightS.size())
            return *leftS.rbegin();
        else
            return ((double)*leftS.rbegin() + *rightS.begin()) / 2.0;
    }
    multiset<int> leftS, rightS;
};

int main()
{
    Solution s;
    vector<int> nums = {1,3,-1,-3,5,3,6,7};
    vector<double> vec = s.medianSlidingWindow(nums, 3);
    for (double d : vec)
        cout << d << " ";
    cout << endl;
    return 0;
}

File: slidingWindowMedian/slidingWindowMedian.java
class Solution {
    public double[] medianSlidingWindow(int[] nums, int k) {
        Comparator<Integer> comparator = (a, b) -> nums[a] != nums[b] ? Integer.compare(nums[a], nums[b]) : a - b;
        TreeSet<Integer> left = new TreeSet<>(comparator.reversed());
        TreeSet<Integer> right = new TreeSet<>(comparator);
        for (int i = 0; i < k; i++) {
            left.add(i);
        }
        balance(left, right);
        double[] res = new double[nums.length - k + 1];
        int resIndex = 0;
        res[resIndex++] = getMedian(k, nums, left, right);
        for (int i = k; i < nums.length; i++) {
            // remove i - k
            if (left.contains(i - k)) left.remove(i - k);
            else right.remove(i - k);
            // add i
            right.add(i);
            left.add(right.pollFirst());
            balance(left, right);
            res[resIndex++] = getMedian(k, nums, left, right);
        }
        return res;
    }
    
    private void balance(TreeSet<Integer> left, TreeSet<Integer> right) {
        while (left.size() > right.size() + 1) {
            right.add(left.pollFirst());
        }
    }
    
    private double getMedian(int k, int[] nums, TreeSet<Integer> left, TreeSet<Integer> right) {
        if (k % 2 == 0) return ((double)nums[left.first()] + nums[right.first()]) / 2;
        else return (double) nums[left.first()];
    }
}
